pkg/integrations/v4/dm/emitter.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/newrelic/infrastructure-agent/internal/feature_flags"
+	"github.com/newrelic/infrastructure-agent/internal/integrations/v4/integration"
+	"github.com/newrelic/infrastructure-agent/pkg/backend/http"
+	"github.com/newrelic/infrastructure-agent/pkg/databind/pkg/data"
+	"github.com/newrelic/infrastructure-agent/pkg/entity"
+	"github.com/newrelic/infrastructure-agent/pkg/integrations/legacy"
+	"github.com/newrelic/infrastructure-agent/pkg/integrations/v4/protocol"
+	"github.com/newrelic/infrastructure-agent/pkg/log"
+)
+
+var (
+	// Errors
+	ProtocolV4NotEnabledErr = errors.New("integration protocol version 4 is not enabled")
+	NoContentToParseErr     = errors.New("no content to parse")
+
+	// internal
+	elog = log.WithComponent("DimensionalMetricsEmitter")
+)
+
+const (
+	nrEntityId = "nr.entity.id"
+)
+
+type Agent interface {
+	GetContext() agent.AgentContext
+}
+
+type emitter struct {
+	ffRetriever   feature_flags.Retriever
+	metricsSender MetricsSender
+	agentContext  agent.AgentContext
+	idProvider    idProviderInterface
+}
+
+type Emitter interface {
+	Send(
+		metadata integration.Definition,
+		extraLabels data.Map,
+		entityRewrite []data.EntityRewrite,
+		integrationJSON []byte) error
+}
+
+func NewEmitter(
+	agentContext agent.AgentContext,
+	dmSender MetricsSender,
+	ffRetriever feature_flags.Retriever,
+	idProvider idProviderInterface) Emitter {
+
+	return &emitter{
+		agentContext:  agentContext,
+		metricsSender: dmSender,
+		ffRetriever:   ffRetriever,
+		idProvider:    idProvider,
+	}
+}
+
+func (e *emitter) Send(
+	metadata integration.Definition,
+	extraLabels data.Map,
+	entityRewrite []data.EntityRewrite,
+	integrationJSON []byte) error {
+
+	pluginDataV4, err := ParsePayloadV4(integrationJSON, e.ffRetriever)
+	if err != nil {
+		elog.WithError(err).WithField("output", string(integrationJSON)).Warn("can't parse v4 integration output")
+		return err
+	}
+
+	return e.process(metadata, extraLabels, entityRewrite, pluginDataV4)
+}
+
+func (e *emitter) process(
+	metadata integration.Definition,
+	extraLabels data.Map,
+	entityRewrite []data.EntityRewrite,
+	integrationData protocol.DataV4) (err error) {
+
+	pluginId := metadata.PluginID(integrationData.Integration.Name)
+	plugin := agent.NewExternalPluginCommon(pluginId, e.agentContext, metadata.Name)
+	labels, extraAnnotations := metadata.LabelsAndExtraAnnotations(extraLabels)
+
+	var entities []protocol.Entity
+	datasetsByEntityName := make(map[string]protocol.Dataset, len(integrationData.DataSets))
+	// Collect All entities
+	for i := range integrationData.DataSets {
+		entities = append(entities, integrationData.DataSets[i].Entity)
+		datasetsByEntityName[integrationData.DataSets[i].Entity.Name] = integrationData.DataSets[i]
+	}
+
+	agentShortName, err := e.agentContext.IDLookup().AgentShortEntityName()
+	if err != nil {
+		return wrapError(fmt.Errorf("error renaming entity: %s", err.Error()), len(integrationData.DataSets))
```

Comment:
`error renaming entity`? this does not seem to be the error

Reply:
we need this short name to rename entity a few rows below. We can change it into 

`"error retrieving short agent entity name: %s"`